### ISSUE TITLE
Login System

### ISSUE THREAD

**Issue by ChickenNuggers** [OpenWebHost/openwebhost.ml#3]
Originally I thought that it would be a smart idea to keep the app confined within a single page, but I do not believe this would be a good idea anymore. Instead, a system based on submitting to / with POST data about logging should be considered.

Open for debate on how to do this.

**Comment by drGrove** [comment]
So there are two ways you can handle this, either sessions or cookie (if you're just going to use flask). If you want others to be able to handle logins and access your api via CORS, you'd want to use something like JWT. But then also create a separation of your UI from you API. Either way is valid. It mostly just comes down to either wanting a contained application or a front-end that takes advantage of the API you're building.

**Comment by ChickenNuggers** [comment]
I was debating on whether it would be possible to pass a diff-like system through JSON to the server once the user is authenticated. Is it possible that we can rewrite the interface to include the password when submitting a modification request?

**Comment by drGrove** [comment]
@ChickenNuggers, what do you mean by diff-like?

**Comment by ChickenNuggers** [comment]
Python has a dict.update() method that updates values based on new ones from the request. As an example, let's say there is a user with these settings:
```json
{
 "branch": "master",
 "keys": []
}
```
Updating keys would be as simple as a request with:
```json
{
 "keys": ["ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCrL6DIvt9zsNI6QDoQsLUsTXEuMAYdliG+oEfbcHxiECChEga/OdeOCQb7bCgk45OOyxUG9HIroP4wKMJ7LpOYZAVoZBjxDmWX76PBADN7cQDWQAmQ4+c4K8HF6hYPERA+kT5/ddkmPL15h4Z0EJUDqUCIM6zWcWmVhKOVpl6oqD8VeKzrVId+gIkJlzRa834FeJ+h2zmzKjdM0VBN8J7jzM1qXw5D95DDX9LU9KvoVfFfRcXXP3AJOlc4pMMCSiUvXz99pDNsIcpYHvP4LOAtElcfG8u+vQu7FDSga8dpGfXmSVsa9xNByG3bzgnNVJprezWuTPmNgPMrWk4srUwF charles@desktop"]
}
```
---
I couldn't think of another way to describe it.

**Comment by drGrove** [comment]
Yeah, that'd work. There are a lot of different ways you can handle this type of thing. But this is pretty standard.

As per the HTTP RFC on PATCH and PUT, This would technically be closer to PATCH, but no one really follows the RFC. 

I've tried to follow, this. http://restcookbook.com/.
It gives a really nice overview, along with links to the appropriate RFCs.

**Comment by ChickenNuggers** [comment]
I would prefer to use `PUT` just to make it compliant, but it is completely able to use `PATCH` with `$.ajax()` if we wish to implement this in the future.

**Comment by ChickenNuggers** [comment]
It looks like `$.post()` is a thing. That would be **easier** to implement but not entirely RFC compliant.

**Comment by drGrove** [comment]
POST tends to be for creating new items in a database or logging in.

**This is a typical flow for a RESTful API**
POST /article
* Creates a new article

GET /article
* Returns all articles

GET /article/:id
* Returns an article

PUT /article/:id
* Updates an article by id.
* Normally requires whole object. But a ton of people tend to do diff

DELETE /article/:id
* Removes the article from the database (or set as inactive)

As you'll tend to find. Most people do not implement PATCH.


On Tue, Oct 27, 2015 at 1:57 PM, Charles Heywood <notifications@github.com>
wrote:

> It looks like $.post() is a thing. That would be *easier* to implement
> but not entirely RFC compliant.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenWebHost/openwebhost.ml/issues/3#issuecomment-151641388>
> .
>



-- 
Danny Grove
DR Grove Software LLC
(321) 348-4932

**Comment by ChickenNuggers** [comment]
>PUT /article/:id
\* Updates an article by id.
\* Normally requires whole object. But a ton of people tend to do diff

I still plan on using POST, but otherwise, it looks like PUT is also a good option.

**Comment by ChickenNuggers** [comment]
[Draft](https://gist.github.com/ChickenNuggers/95121e247dc39814224f) of base protocol.
---
Instead of using a more complex system, users will be able to enter the required information changes and click on "Update", which will send a `POST` to the server with the JSON to patch. This removes the need to store sessions in any way, and users shouldn't need to stay logged in for too long anyways. If they do, they can always create a simple `curl` command to do it for them.

Example `curl` command (untested):
```
curl -F "data=<-" -H "Authentication: Basic ChickenNuggers:$(echo 'Password: ' 1>&2; cat | head -n1)" https://openwebhost.ml < patch.json
```
